pkg/network/server.go

```diff
@@ -138,6 +138,8 @@ type (
 		stateSync StateSync
 
 		log *zap.Logger
+
+		shutdownOnce sync.Once
```

Comment:
Where have you caught the "closed twice" problem? Is it a real reproducible problem?

Reply:
if i run the updated TestServerRegisterPeer  with `defer s.Shutdown()` inside, then i get:

```

=== RUN   TestServerRegisterPeer
    logger.go:130: 2024-02-09T18:23:19.197+0300	INFO	ExtensiblePoolSize is not set or wrong, using default value	{"ExtensiblePoolSize": 20}
    logger.go:130: 2024-02-09T18:23:19.198+0300	INFO	bad AttemptConnPeers configured, using the default value	{"configured": 0, "actual": 20}
    logger.go:130: 2024-02-09T18:23:19.198+0300	INFO	node started	{"blockHeight": 0, "headerHeight": 0}
    logger.go:130: 2024-02-09T18:23:19.198+0300	INFO	node reached synchronized state, starting services
    logger.go:130: 2024-02-09T18:23:19.198+0300	INFO	new peer connected	{"addr": "0.0.0.0:3", "peerCount": 1}
    logger.go:130: 2024-02-09T18:23:19.198+0300	INFO	new peer connected	{"addr": "0.0.0.0:1", "peerCount": 2}
    logger.go:130: 2024-02-09T18:23:19.198+0300	INFO	started protocol	{"addr": "0.0.0.0:3", "userAgent": "fake", "startHeight": 0, "id": 2}
    logger.go:130: 2024-02-09T18:23:19.198+0300	INFO	new peer connected	{"addr": "0.0.0.0:2", "peerCount": 3}
    logger.go:130: 2024-02-09T18:23:19.199+0300	INFO	started protocol	{"addr": "0.0.0.0:1", "userAgent": "fake", "startHeight": 0, "id": 0}
    logger.go:130: 2024-02-09T18:23:19.199+0300	INFO	started protocol	{"addr": "0.0.0.0:2", "userAgent": "fake", "startHeight": 0, "id": 1}
    server_test.go:150: Registered and handshaked peer: 2
    server_test.go:153: Attempted to register extra peer: 3
    server_test.go:150: Registered and handshaked peer: 1
peer dropped: max peers reached
    logger.go:130: 2024-02-09T18:23:19.199+0300	WARN	peer disconnected	{"addr": "0.0.0.0:3", "error": "max peers reached", "peerCount": 2}
    logger.go:130: 2024-02-09T18:23:19.209+0300	WARN	peer disconnected	{"addr": "0.0.0.0:1", "error": "identical node id", "peerCount": 1}
    logger.go:130: 2024-02-09T18:23:19.259+0300	INFO	shutting down server	{"peers": 1}
peer dropped: server shutdown
    logger.go:130: 2024-02-09T18:23:19.259+0300	WARN	peer disconnected	{"addr": "0.0.0.0:2", "error": "server shutdown", "peerCount": 0}
    logger.go:130: 2024-02-09T18:23:19.259+0300	INFO	shutting down server	{"peers": 0}
--- FAIL: TestServerRegisterPeer (0.06s)
panic: closed twice [recovered]
	panic: closed twice

goroutine 60 [running]:
testing.tRunner.func1.2({0x104c40b20, 0x104d0ec90})
	/opt/homebrew/opt/go/libexec/src/testing/testing.go:1545 +0x1c8
testing.tRunner.func1()
	/opt/homebrew/opt/go/libexec/src/testing/testing.go:1548 +0x360
panic({0x104c40b20?, 0x104d0ec90?})
	/opt/homebrew/opt/go/libexec/src/runtime/panic.go:914 +0x218
github.com/nspcc-dev/neo-go/pkg/network.(*fakeTransp).Close(0x14000159e08?)
	/Users/ekaterinapavlova/Workplace/neo-go/pkg/network/discovery_test.go:83 +0x54
github.com/nspcc-dev/neo-go/pkg/network.(*Server).Shutdown(0x14000343400)
	/Users/ekaterinapavlova/Workplace/neo-go/pkg/network/server.go:299 +0x104
github.com/nspcc-dev/neo-go/pkg/network.startWithCleanup.func1()
	/Users/ekaterinapavlova/Workplace/neo-go/pkg/network/server_test.go:408 +0x20
testing.(*common).Cleanup.func1()
	/opt/homebrew/opt/go/libexec/src/testing/testing.go:1169 +0x110
testing.(*common).runCleanup(0x1400032c340, 0x14000159d80?)
	/opt/homebrew/opt/go/libexec/src/testing/testing.go:1347 +0xd8
testing.tRunner.func2()
	/opt/homebrew/opt/go/libexec/src/testing/testing.go:1589 +0x2c
testing.tRunner(0x1400032c340, 0x104d0c5d0)
	/opt/homebrew/opt/go/libexec/src/testing/testing.go:1601 +0x114
created by testing.(*T).Run in goroutine 1
	/opt/homebrew/opt/go/libexec/src/testing/testing.go:1648 +0x33c


Process finished with the exit code 1
```

Reply:
in TestTryInitStateSync i got a similar situation

Reply:
OK, so the fix itself is valid. Attach this description and logs to the corresponding commit.